src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,53 +1,105 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List. Дотриматися основних вимог щодо
  * реалізації ArrayList (default capacity, newCapacity...)</p>
  */
 public class ArrayList<T> implements List<T> {
+    private static int DEFAULT_CAPACITY = 10;
+    private T[] arrayElements;
+    private int size;
+
+    public ArrayList() {
+        arrayElements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public T[] newArrayOfElements() {
+        T[] newArrayElements = arrayElements;
+        arrayElements = (T[]) new Object[arrayElements.length + (arrayElements.length >> 1)];
+        System.arraycopy(newArrayElements, 0, arrayElements, 0, newArrayElements.length);
+        return arrayElements;
+    }
+
+    private boolean indexChecking(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayIndexOutOfBoundsException("The index is not correct");
+        }
+        return true;
+    }
 
     @Override
     public void add(T value) {
-
+        if (size > arrayElements.length - 1) {
+            newArrayOfElements();
+        }
+        arrayElements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException("The index is not correct");
+        }
+        if (size > arrayElements.length) {
+            newArrayOfElements();
+        }
+        System.arraycopy(arrayElements, index, arrayElements, index + 1, size - index);
+        arrayElements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (size > arrayElements.length) {
+            newArrayOfElements();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            arrayElements[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexChecking(index);
+        return arrayElements[index];
     }
 
     @Override
     public void set(T value, int index) {
+        indexChecking(index);
+        arrayElements[index] = value;
 
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexChecking(index);
+        T temp = arrayElements[index];
+        System.arraycopy(arrayElements, index + 1, arrayElements, index, size - index);
```

Comment:
```suggestion
        System.arraycopy(arrayElements, index + 1, arrayElements, index, size - index - 1);
```
we should move 1 element less because we are removing here

Reply:
Done